co2calculator/calculate.py

```diff
@@ -21,78 +24,149 @@
+
+
+@enum.unique
+class TransportationMode(str, enum.Enum):
+    """Available transportation modes"""
+
+    CAR = "car"
+    BUS = "bus"
+    TRAIN = "train"
+    PLANE = "plane"
+    FERRY = "ferry"
+
+
+class EmissionRequest(BaseModel):
+    """Generic model for emission calculation request"""
+
+    transportation_mode: TransportationMode
+    distance: Kilometer
+
+
+class Stop(BaseModel):  # Could be renamed to "waypoint"?
+    """Model for a waypoint of a trip"""
+
+    address: str
+    locality: str
+    country: str
+
+
+class StopList(BaseModel):
+    """Model for a list of waypoints"""
+
+    __root__: List[Stop]
+
+    def __iter__(self):
+        return iter(self.__root__)
+
+    def __getitem__(self, item):
+        return self.__root__[item]
+
+
+class DistanceRequest(BaseModel):
+    """Request-model to calculate distances"""
+
+    stops: StopList
+    transportation_mode: TransportationMode
+
+
+def trip_direct(request: DistanceRequest) -> Kilometer:
+    """Function to calculate the distance of a trip as a direct route between locations
+    :param request: validated input data
+    :return: Distance of the trip
+    :rtype: Kilometer
+
     """
-    # NOTE: Tests fail for 'cng'  as `fuel_type` (IndexError)
+    coords = []
+    for loc in DistanceRequest.stops:
+        loc_name, loc_country, loc_coords, _ = geocoding_structured(loc.dict())
+        coords.append(loc_coords)
+    distance = get_route(coords, "driving-car")
+    return distance
 
-    transport_mode = "car"
 
-    # Set default values
-    if passengers is None:
-        passengers = 1
-        warnings.warn(
-            f"Number of car passengers was not provided. Using default value: '{passengers}'"
-        )
-    if size is None:
-        size = "average"
-        warnings.warn(f"Size of car was not provided. Using default value: '{size}'")
-    if fuel_type is None:
-        fuel_type = "average"
-        warnings.warn(
-            f"Car fuel type was not provided. Using default value: '{fuel_type}'"
-        )
-    # Check if distance of stops provided
-    if distance is None and stops is None:
-        raise ValueError(
-            "Travel parameters missing. Please provide either the distance in km or a list of"
-            "dictionaries of travelled locations."
+def trip_detour(request: DistanceRequest) -> Kilometer:
+    """
+    Function to calculate the distance of a trip using the distance as the crow flies and detour coefficients/constants
+
+    :param request: validated input data
+    :return: Distance in kilometer
+    :rtype: Kilometer
+    """
+    distance = 0
+    coords = []
+    for loc in DistanceRequest.stops:
+        loc_name, loc_country, loc_coords, _ = geocoding_structured(loc.dict())
+        coords.append(loc_coords)
+    for i in range(0, len(coords) - 1):
+        # compute great circle distance between locations
+        distance += haversine(
+            coords[i][1], coords[i][0], coords[i + 1][1], coords[i + 1][0]
```

Comment:
```python
class Coordinate(BaseModel):
    lat: float
    lng: float
```

.. would be a great next mover!

Without refactoring of `haversine()`, the code would look like:
```python
distance += haversine(coords[i].lat, coords[i].lng, coords[i+1].lat, coords[i+1].lng)
```

With a little refactoring:
```python
distance += haversine(coords[i], coords[i+1])
```